browser/src/control/jsdialog/Util.KeyboardGridNavigation.ts

```diff
@@ -0,0 +1,166 @@
+					return row === currentRow && column === targetColumn;
+				})
+			) {
+				targetColumn = 0; // Start from the first column
+			}
+		} else {
+			targetColumn--;
+			// If it's the first element in the row and trying to move previous, cycle to the last in the same row
+			if (targetColumn < 0) {
+				targetColumn =
+					focusableElements.filter((el) => {
+						const [row] = getRowColumn(el);
+						return row === currentRow;
+					}).length - 1; // Move to the last column in the same row
+			}
+		}
+	} else if (axis === 'vertical') {
+		if (direction === 'next') {
+			targetRow++;
+		} else {
+			targetRow--;
+		}
+	}
+
+	// Find the target element based on the calculated row and column
+	const targetElement = focusableElements.find((el) => {
+		const [row, column] = getRowColumn(el);
+		return row === targetRow && column === targetColumn;
+	});
+
+	if (!targetElement && axis === 'vertical') {
+		if (direction === 'next') {
+			// Start from the next sibling of the parent container
+			const nextFocusableElement = findFocusableElement(
+				nextElement as HTMLElement,
+				'next',
+				isFocusable,
+			);
+			if (nextFocusableElement) {
+				nextFocusableElement.focus();
+			}
+		} else if (direction === 'previous') {
+			// Start from the previous sibling of the parent container
+			const previousFocusableElement = findFocusableElement(
+				nextElement as HTMLElement,
+				'previous',
+				isFocusable,
+			);
+			if (previousFocusableElement) {
+				previousFocusableElement.focus();
+			}
+		}
+	}
+
+	if (targetElement) {
+		targetElement.focus();
+	}
+}
+
+function findFocusableElement(
+	element: HTMLElement,
+	direction: 'next' | 'previous',
+	isFocusable: (el: HTMLElement) => boolean,
+): HTMLElement | null {
+	// Check the current element if it is focusable
+	if (isFocusable(element)) return element;
+
+	// Check if sibling is focusable or contains focusable elements
+	const focusableInSibling = findFocusableWithin(
+		element as HTMLElement,
+		direction,
+	);
+	if (focusableInSibling) return focusableInSibling;
+
+	// Depending on the direction, find the next or previous sibling
+	const sibling: Element =
+		direction === 'next'
+			? element.nextElementSibling
+			: element.previousElementSibling;
+
+	if (sibling) {
+		// Recursively check the next or previous sibling of the current sibling
+		return findFocusableElement(sibling as HTMLElement, direction, isFocusable);
+	}
+
+	return null;
+}
+
+// Helper function to find the first focusable element within an element
+function findFocusableWithin(
+	element: HTMLElement,
+	direction: string,
+): HTMLElement | null {
+	const focusableElements = Array.from(element.querySelectorAll('*'));
+	return direction === 'next'
+		? (focusableElements.find(isFocusable) as HTMLElement | null)
+		: (focusableElements.reverse().find(isFocusable) as HTMLElement | null);
+}
+
+// Utility function to check if an element is focusable
+function isFocusable(element: HTMLElement) {
```

Comment:
we already have util like that, please reuse: https://github.com/CollaboraOnline/online/blob/f55c1a1403fcd350e81fdb98bd6c3a6b05cfda2b/browser/src/control/jsdialog/Util.FocusCycle.js#L18

Reply:
yes we have but it returns list of focusable element so i think if there is a function which can help to find the current element is focusable or not can be more helpful. But yes, i will this into `util.FocusCycle.js`

Reply:
I see at many places we are using same function but then after to get first focusable element we do like
```
FocusableElements[0].focus

```

This seems not okay because we are adding many code duplication here so by this `isFocusable` function we can directly get our first focusable element.

What you say @eszkadev on this ?

Reply:
yes, please try to reuse as much as possible. If we need: `getFirstFocusableElement()` we casn add it.
best to keep everything in the util file you mentioned so everything is in one place

Reply:
sure @eszkadev now all function are in Util.FocusCycle.js